CppSamples/EditData/EditGeometriesWithProgrammaticReticleTool/EditGeometriesWithProgrammaticReticleTool.cpp

```diff
@@ -0,0 +1,535 @@
+      case GeometryEditorMode::MultipointMode:
+        m_geometryEditor->start(GeometryType::Multipoint);
+        break;
+      case GeometryEditorMode::PolylineMode:
+        m_geometryEditor->start(GeometryType::Polyline);
+        break;
+      case GeometryEditorMode::PolygonMode:
+        m_geometryEditor->start(GeometryType::Polygon);
+        break;
+      default:
+        break;
+    }
+    emit geometryEditorStartedChanged();
+    setMultifunctionButtonState();
+    return;
+  }
+
+  if (m_vertexCreationAllowed)
+  {
+    switch (m_reticleState)
+    {
+      case ReticleState::Default:
+      case ReticleState::PickedUp:
+        m_reticleTool->placeElementAtReticle();
+        break;
+      case ReticleState::HoveringVertex:
+      case ReticleState::HoveringMidVertex:
+        m_reticleTool->selectElementAtReticle();
+        m_reticleTool->pickUpSelectedElement();
+        break;
+    }
+  }
+  else
+  {
+    switch (m_reticleState)
+    {
+      case ReticleState::PickedUp:
+        m_reticleTool->placeElementAtReticle();
+        break;
+      case ReticleState::HoveringVertex:
+        m_reticleTool->selectElementAtReticle();
+        m_reticleTool->pickUpSelectedElement();
+        break;
+      default:
+        // In edit mode, only picked up vertices can be placed and only vertices can be picked up.
+        break;
+    }
+  }
+  setMultifunctionButtonState();
+}
+
+void EditGeometriesWithProgrammaticReticleTool::saveEdits()
+{
+  // Stop the geometry editor and get the resulting geometry
+  const auto geometry = m_geometryEditor->stop();
+  emit geometryEditorStartedChanged();
+
+  if (!geometry.isEmpty())
+  {
+    if (m_editingGraphic)
+    {
+      // Update the geometry of the graphic being edited and make it visible again
+      m_editingGraphic->setGeometry(geometry);
+      m_editingGraphic->setVisible(true);
+      m_editingGraphic = nullptr;
+    }
+    else
+    {
+      Symbol* geometrySymbol = getSymbol(geometry.geometryType());
+      if (geometrySymbol)
+      {
+        m_graphicsOverlay->graphics()->append(new Graphic(geometry, geometrySymbol, this));
+      }
+    }
+  }
+  resetFromEditingSession();
+}
+
+Symbol* EditGeometriesWithProgrammaticReticleTool::getSymbol(GeometryType type) const
+{
+  switch (type)
+  {
+    case GeometryType::Point:
+      return m_pointSymbol;
+    case GeometryType::Multipoint:
+      return m_multiPointSymbol;
+    case GeometryType::Polyline:
+      return m_lineSymbol;
+    case GeometryType::Polygon:
+      return m_polygonSymbol;
+    default:
+      return nullptr;
+  }
+}
+
+void EditGeometriesWithProgrammaticReticleTool::discardEdits()
+{
+  m_geometryEditor->stop();
+  resetFromEditingSession();
+}
+
+void EditGeometriesWithProgrammaticReticleTool::resetFromEditingSession()
+{
+  // Reset the selected graphic
+  if (m_editingGraphic)
+  {
+    m_editingGraphic->setSelected(false);
+    m_editingGraphic->setVisible(true);
+  }
+  m_editingGraphic = nullptr;
```

Comment:
Should we just delete `m_editingGraphic`? Why bother to unselect and set it to visible if we're setting it to null?